django/contrib/staticfiles/management/commands/collectstatic.py

```diff
@@ -279,7 +279,7 @@ def delete_file(self, path, prefixed_path, source_storage):
                     # Skip the file if the source file is younger
                     # Avoid sub-second precision (see #14665, #19540)
                     if (target_last_modified.replace(microsecond=0) >= source_last_modified.replace(microsecond=0) and
-                            full_path and not (self.symlink ^ os.path.islink(full_path))):
+                            not full_path or full_path and not (self.symlink ^ os.path.islink(full_path))):
```

Comment:
I think that's an incorrect grouping. in Django 1.9, the condition looks like this:
```
if (target_last_modified.replace(microsecond=0)
        >= source_last_modified.replace(microsecond=0)):
    if not ((self.symlink and full_path
             and not os.path.islink(full_path)) or
            (not self.symlink and full_path
             and os.path.islink(full_path))):
```
It looks like `A and (not B or (B and not C))` might be the correct grouping but a comment would really help explain what exactly the logic is doing and it seems that tests are still missing for incorrect groupings.